# Aufgabenbearbeitung I - Datenanalyse (Lösungsansatz)
<br>
<br>
Hinweise: https://pandas.pydata.org/docs/user_guide/index.html

In [1]:
import pandas as pd  #Einfügen der Pandas-Bibliothek über die Abkürzung "pd"
import datetime      #um die Datentypen Datum und Zeit nutzen zu können

In [2]:
df = pd.read_csv("C:/Users/Rocky Telatko/Documents/RT/Python/IoT-Workshop_Woche/ProdTag_12_05_2021.csv", 
                 sep = ";")   #Einlesen der csv-Daten
df.head(n=10)   #Ausgabe der ersten 10 Zeilen des Datensatzes

Produktvariante  Prozessschritt  Auftragsnummer  \
0             8010              10            1329   
1             8010              20            1329   
2             8010              30            1329   
3             8010              40            1329   
4             8010              50            1329   
5             8010              60            1329   
6             8010              10            1329   
7             8010              20            1329   
8             8010              30            1329   
9             8010              40            1329   

                   Prozessbeschreibung  Prozessnummer  Startdatum Startzeit  \
0  release a defined part on stopper 1            212  12.05.2021  10:02:25   
1               feed part from magazin            200  12.05.2021  10:03:07   
2    assemble a PCB with one fuse left            302  12.05.2021  10:04:45   
3               feed part from magazin            200  12.05.2021  10:07:27   
4         pressing for time in seconds            110  12.05.2021  10:07:43   
5          store a part from stopper 2            211  12.05.2021  10:09:49   
6  release a defined part on stopper 1            212  12.05.2021  10:02:43   
7               feed part from magazin            200  12.05.2021  10:03:23   
8    assemble a PCB with one fuse left            302  12.05.2021  10:06:35   
9               feed part from magazin            200  12.05.2021  10:09:28   

     Enddatum   Endzeit  Arbeitsstation  Transportzeit  
0  12.05.2021  10:02:40               5              0  
1  12.05.2021  10:03:09              11             27  
2  12.05.2021  10:06:28               8              6  
3  12.05.2021  10:07:30               6             63  
4  12.05.2021  10:07:45               7             14  
5  12.05.2021  10:10:01               5           1126  
6  12.05.2021  10:02:56               5              0  
7  12.05.2021  10:03:25              11             27  
8  12.05.2021  10:08:18               8              6  
9  12.05.2021  10:09:30               6             63

In [3]:
df.info()   #Ausgabe der Datensatzstruktur

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1314 entries, 0 to 1313
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Produktvariante      1314 non-null   int64 
 1   Prozessschritt       1314 non-null   int64 
 2   Auftragsnummer       1314 non-null   int64 
 3   Prozessbeschreibung  1314 non-null   object
 4   Prozessnummer        1314 non-null   int64 
 5   Startdatum           1307 non-null   object
 6   Startzeit            1307 non-null   object
 7   Enddatum             1307 non-null   object
 8   Endzeit              1307 non-null   object
 9   Arbeitsstation       1314 non-null   int64 
 10  Transportzeit        1314 non-null   int64 
dtypes: int64(6), object(5)
memory usage: 113.0+ KB


In [4]:
df["Zykluszeit"] = pd.to_datetime(df["Endzeit"]) - pd.to_datetime(df["Startzeit"])   #Berechnung der Zykluszeit je Prozessschritt
df["Zykluszeit"] = df["Zykluszeit"].dt.total_seconds()   #Umwandlung der Zykluszeit in Sekunden
df.head(n=5)   #Ausgabe der ersten 5 Zeilen des Datensatzes

Produktvariante  Prozessschritt  Auftragsnummer  \
0             8010              10            1329   
1             8010              20            1329   
2             8010              30            1329   
3             8010              40            1329   
4             8010              50            1329   

                   Prozessbeschreibung  Prozessnummer  Startdatum Startzeit  \
0  release a defined part on stopper 1            212  12.05.2021  10:02:25   
1               feed part from magazin            200  12.05.2021  10:03:07   
2    assemble a PCB with one fuse left            302  12.05.2021  10:04:45   
3               feed part from magazin            200  12.05.2021  10:07:27   
4         pressing for time in seconds            110  12.05.2021  10:07:43   

     Enddatum   Endzeit  Arbeitsstation  Transportzeit  Zykluszeit  
0  12.05.2021  10:02:40               5              0        15.0  
1  12.05.2021  10:03:09              11             27         2.0  
2  12.05.2021  10:06:28               8              6       103.0  
3  12.05.2021  10:07:30               6             63         3.0  
4  12.05.2021  10:07:45               7             14         2.0

In [5]:
df.dtypes   #Ausgabe der Datentypen je Spalte des Datensatzes

Produktvariante          int64
Prozessschritt           int64
Auftragsnummer           int64
Prozessbeschreibung     object
Prozessnummer            int64
Startdatum              object
Startzeit               object
Enddatum                object
Endzeit                 object
Arbeitsstation           int64
Transportzeit            int64
Zykluszeit             float64
dtype: object

In [6]:
#Transportzeit und Zykluszeit je Arbeitsstation (summiert)
selectAS = df[["Arbeitsstation", "Transportzeit", "Zykluszeit"]]   #Selektieren des Datensatzes nach den beschriebenen Spalten
prozessDauernAS = selectAS.groupby("Arbeitsstation").sum()   #Gruppierung nach den Arbeitsstationen und Aufsummieren der Werte
prozessDauernAS["Gesamtzeit"] = prozessDauernAS["Transportzeit"] + prozessDauernAS["Zykluszeit"]   #Berechnung der Gesamtzeit in extra Spalte
prozessDauernAS   #Ergebnisausgabe

Transportzeit  Zykluszeit  Gesamtzeit
Arbeitsstation                                       
5                      246594      5696.0    252290.0
6                       13797      1090.0     14887.0
7                        3066       620.0      3686.0
8                        1314     23361.0     24675.0
11                       5913       455.0      6368.0

In [7]:
#NaN werden mitunter eigenständig gefiltert: "Operations in general exclude missing data" (das muss aber nicht sein!)
selectAS2 = selectAS.dropna()   #Entfernen von Zeilen mit fehlenden Werten
prozessAnz1 = selectAS.groupby("Arbeitsstation").count()   #Gruppierung nach Arbeitsstationen und Zählen der eingetragenen Werte je Station
prozessAnz2 = selectAS2.groupby("Arbeitsstation").count()   #Gruppierung nach Arbeitsstationen und Zählen der eingetragenen Werte je Station
print(prozessAnz1)   #Ergebnisausgabe
print(prozessAnz2)   #Ergebnisausgabe
#Anhand der Transportzeit ist die Filterung der NaN-Werte zu erkennen, da die Zykluszeit unter Ausschluss der NaN-Werte berechnet wurde 

                Transportzeit  Zykluszeit
Arbeitsstation                           
5                         438         435
6                         219         217
7                         219         217
8                         219         219
11                        219         219
                Transportzeit  Zykluszeit
Arbeitsstation                           
5                         435         435
6                         217         217
7                         217         217
8                         219         219
11                        219         219


In [8]:
meanProzessDauerAS = selectAS.groupby("Arbeitsstation").mean()   #Gruppierung nach den Arbeitsstationen und ermitteln des Durchschnitts über die Werte
meanProzessDauerAS["Gesamtzeit"] = meanProzessDauerAS["Transportzeit"] + meanProzessDauerAS["Zykluszeit"]   #Ermitteln der Gesamtzeit in extra Spalte
meanProzessDauerAS   #Ergebnisausgabe

Transportzeit  Zykluszeit  Gesamtzeit
Arbeitsstation                                       
5                       563.0   13.094253  576.094253
6                        63.0    5.023041   68.023041
7                        14.0    2.857143   16.857143
8                         6.0  106.671233  112.671233
11                       27.0    2.077626   29.077626

In [9]:
#Durchschnittliche Transportzeit und Zykluszeit je Arbeitsstation und Produktvariante
selectAS_PV = df[["Arbeitsstation", "Produktvariante", "Transportzeit", "Zykluszeit"]]   #Selektieren des Datensatzes nach den beschriebenen Spalten
meanProzessDauerAS_PV = selectAS_PV.groupby(["Arbeitsstation", "Produktvariante"]).mean()   ##Gruppierung nach den beschriebenen Spalten und ermitteln des Durchschnitts über die Werte
meanProzessDauerAS_PV["Gesamtzeit"] = meanProzessDauerAS_PV["Transportzeit"] + meanProzessDauerAS_PV["Zykluszeit"]   #Ermitteln der Gesamtzeit in extra Spalte
meanProzessDauerAS_PV   #Ergebnisausgabe

Transportzeit  Zykluszeit  Gesamtzeit
Arbeitsstation Produktvariante                                       
5              8010                     563.0   12.888889  575.888889
               8011                     563.0   13.106195  576.106195
               8110                     563.0   13.063830  576.063830
               8111                     563.0   13.324324  576.324324
6              8010                      63.0   12.482759   75.482759
               8011                      63.0    2.333333   65.333333
               8110                      63.0    2.255319   65.255319
               8111                      63.0    2.309091   65.309091
7              8010                      14.0    2.534483   16.534483
               8011                      14.0    3.736842   17.736842
               8110                      14.0    2.531915   16.531915
               8111                      14.0    2.563636   16.563636
8              8010                       6.0  106.203390  112.203390
               8011                       6.0  120.157895  126.157895
               8110                       6.0   94.000000  100.000000
               8111                       6.0  104.071429  110.071429
11             8010                      27.0    1.898305   28.898305
               8011                      27.0    2.508772   29.508772
               8110                      27.0    1.957447   28.957447
               8111                      27.0    1.928571   28.928571

Auffällig hoher Durchschnittswert bei der Zykluszeit von Arbeitsstation 6 und Produktvariante 8010!?

In [10]:
selectAS_PV1 = df[["Arbeitsstation", "Produktvariante", "Zykluszeit"]]   #Selektieren des Datensatzes nach den beschriebenen Spalten
station6 = selectAS_PV1[selectAS_PV1["Arbeitsstation"] == 6]   #Filtern nach Arbeitsstation 6 (Stapelmagazin Oberschale)
prozessDauerAS_PVanz = station6.groupby(["Arbeitsstation", "Produktvariante", "Zykluszeit"]).count()   #Gruppierung nach den beschriebenen Spalten und Ermitteln der vorkommenden Werte
prozessDauerAS_PVanz   #Ergebnisausgabe

Empty DataFrame
Columns: []
Index: [(6, 8010, 2.0), (6, 8010, 3.0), (6, 8010, 594.0), (6, 8011, 2.0), (6, 8011, 3.0), (6, 8110, 2.0), (6, 8110, 3.0), (6, 8111, 2.0), (6, 8111, 3.0)]

In [11]:
df.loc[(df['Arbeitsstation'] == 6) & (df['Zykluszeit'] == 594.0)]   #Ausgabe des Maximalwertes bei der Zykluszeit von Station 6, um zu prüfen, wie oft dieser Wert vorkommt
#wahrscheinliche Ursache: fehlendes Material oder Defekt

Produktvariante  Prozessschritt  Auftragsnummer     Prozessbeschreibung  \
1083             8010              40            1360  feed part from magazin   

      Prozessnummer  Startdatum Startzeit    Enddatum   Endzeit  \
1083            200  12.05.2021  19:26:39  12.05.2021  19:36:33   

      Arbeitsstation  Transportzeit  Zykluszeit  
1083               6             63       594.0

In [12]:
selectAS_PV.groupby(["Arbeitsstation", "Produktvariante"]).describe()   ##Gruppierung nach den beschriebenen Spalten und zusammenfassen der Werte zu statistischen Größen

Transportzeit                                 \
                                       count   mean         std   min   25%   
Arbeitsstation Produktvariante                                                
5              8010                    118.0  563.0  565.400864   0.0   0.0   
               8011                    114.0  563.0  565.485663   0.0   0.0   
               8110                     94.0  563.0  566.018788   0.0   0.0   
               8111                    112.0  563.0  565.530350   0.0   0.0   
6              8010                     59.0   63.0    0.000000  63.0  63.0   
               8011                     57.0   63.0    0.000000  63.0  63.0   
               8110                     47.0   63.0    0.000000  63.0  63.0   
               8111                     56.0   63.0    0.000000  63.0  63.0   
7              8010                     59.0   14.0    0.000000  14.0  14.0   
               8011                     57.0   14.0    0.000000  14.0  14.0   
               8110                     47.0   14.0    0.000000  14.0  14.0   
               8111                     56.0   14.0    0.000000  14.0  14.0   
8              8010                     59.0    6.0    0.000000   6.0   6.0   
               8011                     57.0    6.0    0.000000   6.0   6.0   
               8110                     47.0    6.0    0.000000   6.0   6.0   
               8111                     56.0    6.0    0.000000   6.0   6.0   
11             8010                     59.0   27.0    0.000000  27.0  27.0   
               8011                     57.0   27.0    0.000000  27.0  27.0   
               8110                     47.0   27.0    0.000000  27.0  27.0   
               8111                     56.0   27.0    0.000000  27.0  27.0   

                                                      Zykluszeit              \
                                  50%     75%     max      count        mean   
Arbeitsstation Produktvariante                                                 
5              8010             563.0  1126.0  1126.0      117.0   12.888889   
               8011             563.0  1126.0  1126.0      113.0   13.106195   
               8110             563.0  1126.0  1126.0       94.0   13.063830   
               8111             563.0  1126.0  1126.0      111.0   13.324324   
6              8010              63.0    63.0    63.0       58.0   12.482759   
               8011              63.0    63.0    63.0       57.0    2.333333   
               8110              63.0    63.0    63.0       47.0    2.255319   
               8111              63.0    63.0    63.0       55.0    2.309091   
7              8010              14.0    14.0    14.0       58.0    2.534483   
               8011              14.0    14.0    14.0       57.0    3.736842   
               8110              14.0    14.0    14.0       47.0    2.531915   
               8111              14.0    14.0    14.0       55.0    2.563636   
8              8010               6.0     6.0     6.0       59.0  106.203390   
               8011               6.0     6.0     6.0       57.0  120.157895   
               8110               6.0     6.0     6.0       47.0   94.000000   
               8111               6.0     6.0     6.0       56.0  104.071429   
11             8010              27.0    27.0    27.0       59.0    1.898305   
               8011              27.0    27.0    27.0       57.0    2.508772   
               8110              27.0    27.0    27.0       47.0    1.957447   
               8111              27.0    27.0    27.0       56.0    1.928571   

                                                                              
                                      std    min    25%    50%    75%    max  
Arbeitsstation Produktvariante                                                
5              8010              1.127843   11.0   12.0   13.0   14.0   16.0  
               8011              1.298095   11.0   12.0 

#### Auswertung der statistischen Werte

Anhand der Anzahl der Transporte (bei AS 5 jeweils zwei Prozessschritte je Werkstück) und der Anzahl der Zykluszeiten lassen sich Rückschlüsse ziehen, an welcher Stelle es Probleme (Störungen) gab --> bei Werkstück 8010 und 8111 bei bzw. nach AS 8, bei Werkstück 8011 bei bzw. nach AS 7 (Abfolge AS 5-11-8-6-7-5).

Die Standardabweichung bei der Transportzeit der AS 5 ist ungewöhnlich. Die Ursache ist, dass die erste erkannte Transportzeit parallel zum ersten Prozessschritt erfasst wird und damit, im Gegensatz zum zweiten Prozessschritt der AS, einen Nullwert darstellt.

Der Maximalwert der Zykluszeit bei AS 7 und Werkstück 8011 (siehe auch die hohe Standardabweichung), neben dem von AS 6 bei Werkstück 8010, signalisiert ebenfalls ein ungeplantes Ereignis bzw. ungeplante Ereignisse. Um zu prüfen, ob es sich um ein Ereignis oder mehrere handelt, wird eine Filterung durchgeführt (siehe Befehl unten).

Es ist anhand der Standardabweichung erkennbar, dass die Engpassstation AS 8 den instabilsten Prozess darstellt --> woran liegt das? Zur Beantwortung der Frage oder dem Erarbeiten weiterer Fragen wären weitere Daten hilfreich!
Ebenfalls auffällig sind die Minimalwerte bei AS 8 bei Werkstück 8010 und 8111, welche ebenfalls auf ungeplante Ereignisse (Störungen) hinweisen (siehe Befehl unten).

Der Maximalwert der Zykluszeit bei AS 11 und Werkstück 8011 signalisiert ein weiteres ungeplantes Ereignis bzw. ungeplante Ereignisse. Um zu prüfen, ob es sich um ein Ereignis oder mehrere handelt, wird eine Filterung durchgeführt (siehe Befehl unten).

**Fazit: Fokus muss auf AS 8 gelegt werden!**

In [26]:
df.loc[(df['Arbeitsstation'] == 7) & (df['Zykluszeit'] >= 10.0)]   #Ausgabe der Zeilen mit überdurchschnittlichen Werten bei der Zykluszeit von Station 7

Produktvariante  Prozessschritt  Auftragsnummer  \
340              8110              50            1335   
520              8111              50            1337   
1144             8011              50            1361   

               Prozessbeschreibung  Prozessnummer  Startdatum Startzeit  \
340   pressing for time in seconds            110  12.05.2021  12:30:40   
520   pressing for time in seconds            110  12.05.2021  13:23:45   
1144  pressing for time in seconds            110  12.05.2021  19:52:19   

        Enddatum   Endzeit  Arbeitsstation  Transportzeit  Zykluszeit  
340   12.05.2021  12:30:50               7             14        10.0  
520   12.05.2021  13:23:56               7             14        11.0  
1144  12.05.2021  19:53:41               7             14        82.0

wahrscheinliche Ursache bei 8011: Defekt

In [28]:
df.loc[(df['Arbeitsstation'] == 11) & (df['Zykluszeit'] >= 5.0)]   #Ausgabe der Zeilen mit überdurchschnittlichen Werten bei der Zykluszeit von Station 11

Produktvariante  Prozessschritt  Auftragsnummer     Prozessbeschreibung  \
1219             8011              20            1362  feed part from magazin   

      Prozessnummer  Startdatum Startzeit    Enddatum   Endzeit  \
1219            200  12.05.2021  20:06:06  12.05.2021  20:06:41   

      Arbeitsstation  Transportzeit  Zykluszeit  
1219              11             27        35.0

wahrscheinliche Ursache: fehlendes Material oder Defekt

In [14]:
'''
Prüfen, wie sehr die Zeiten der Produktvariante 8010 an Arbeitsstation 8 schwankt - sind einzelne oder mehrere Ereignisse betroffen?
Werkstück 8010 hat Durchschnittswert bei 106 Sekunden (Median = 103s)
'''
df.loc[(df['Arbeitsstation'] == 8) & (df['Produktvariante'] == 8010) & (df['Zykluszeit'] >= 110.0)]   #Ausgabe der Zeilen mit überdurchschnittlichen Werten bei der Zykluszeit von Station 8 bei Produktvariante 8010

Produktvariante  Prozessschritt  Auftragsnummer  \
122              8010              30            1333   
302              8010              30            1335   
392              8010              30            1337   
451              8010              30            1339   
457              8010              30            1339   
842              8010              30            1356   
962              8010              30            1358   
986              8010              30            1358   
1202             8010              30            1362   

                    Prozessbeschreibung  Prozessnummer  Startdatum Startzeit  \
122   assemble a PCB with one fuse left            302  12.05.2021  11:33:44   
302   assemble a PCB with one fuse left            302  12.05.2021  12:16:08   
392   assemble a PCB with one fuse left            302  12.05.2021  13:01:38   
451   assemble a PCB with one fuse left            302  12.05.2021  13:40:46   
457   assemble a PCB with one fuse left            302  12.05.2021  13:42:59   
842   assemble a PCB with one fuse left            302  12.05.2021  18:08:48   
962   assemble a PCB with one fuse left            302  12.05.2021  18:46:26   
986   assemble a PCB with one fuse left            302  12.05.2021  19:03:47   
1202  assemble a PCB with one fuse left            302  12.05.2021  20:12:36   

        Enddatum   Endzeit  Arbeitsstation  Transportzeit  Zykluszeit  
122   12.05.2021  11:35:52               8              6       128.0  
302   12.05.2021  12:18:21               8              6       133.0  
392   12.05.2021  13:03:48               8              6       130.0  
451   12.05.2021  13:42:54               8              6       128.0  
457   12.05.2021  13:45:36               8              6       157.0  
842   12.05.2021  18:11:05               8              6       137.0  
962   12.05.2021  18:48:43               8              6       137.0  
986   12.05.2021  19:05:57               8              6       130.0  
1202  12.05.2021  20:14:53               8              6       137.0

Das Ergebnis verdeutlicht den Handlungsbedarf aufgrund der Instabilität der Vorgänge!

In [15]:
df.loc[(df['Arbeitsstation'] == 8) & (df['Produktvariante'] == 8010) & (df['Zykluszeit'] <= 100.0)]   #Ausgabe der Zeilen mit unterdurchschnittlichen Werten bei der Zykluszeit von Station 8 für Produktvariante 8010

Produktvariante  Prozessschritt  Auftragsnummer  \
1082             8010              30            1360   
1094             8010              30            1360   

                    Prozessbeschreibung  Prozessnummer  Startdatum Startzeit  \
1082  assemble a PCB with one fuse left            302  12.05.2021  19:25:23   
1094  assemble a PCB with one fuse left            302  12.05.2021  19:26:23   

        Enddatum   Endzeit  Arbeitsstation  Transportzeit  Zykluszeit  
1082  12.05.2021  19:25:50               8              6        27.0  
1094  12.05.2021  19:27:37               8              6        74.0

Vermutlich handelte es sich hier um eine Störung, wodurch eine Bearbeitung unterbrochen und als zwei separate Bearbeitungen erfasst wurde.